TDD Notes (#25)
paImer/notes: lessons/february/fibonacciTest.md

```java
class Fib {
    int fib(int n) {
        return -1;
    }
}
```

### Refactor III

This does the job, and is already fairly simple, so let's skip this step for now (but only after careful consideration)

### Red IV

Time to make another failing test.
We have already discussed that, as per our definition, `F(1)` should return `1`.

Here we will add another assertion (not an entire test) for brevity.
It is preferable to write different tests when testing different inputs, due to the way JUnit works - a assertion will prevent any of the assertions within the same test from running, depriving you of potentially useful information.
Multiple tests removes this concern.

```java
Assert.assertEquals(1, new Fib().fib(1));
```

### Green IV

It might make sense here to split up our inputs into two camps: those less than 1, and those greater than or equal to 1.
As we are dealing only with integers, this covers all possible inputs.

```java
class Fib {
    int fib(int n) {
        if (n < 1) {
            return -1;
        }
        return 1;
    }
}
```

All tests  pass!

### Refactor IV

> Let's extract a method called `checkFib()` from `testFib()` to improve readability:

Reviewer: Let's also note that we're statically importing `Assert.assertEquals` for readability.

If we try the next logical step `checkFib(1, 2)`, we notice that this assertion passes.
This is not a failing test, and therefore we ignore it and look for the next most degenerate case that results in a failure.

Indeed, `checkFib(2, 3)` does exactly this.

### Green V

One simple case that passes:

```java
class Fib {
    int fib(int n) {
        if (n < 1) {
            return -1;
        }
        if (n == 1) {
            return 1;
        }
        if (n == 2) {
            return 1;
        }
        return 2;
    }
}
```

### Refactor V

Having two inputs return the same value can be tided up a bit.
Let's use the `||` operator to fix this.

```java
class Fib {
    int fib(int n) {
        if (n < 1) {
            return -1;
        }
        if (n == 1 || n == 2) {
            return 1;
        }
        return 2;
    }
}
```

### Red VI

Let's check `F(4) = 3` with `checkFib(3, 4)`.

### Green VI

```java
class Fib {
    int fib(int n) {
        if (n < 1) {
            return -1;
        }
        if (n == 1 || n == 2) {
            return 1;
        }
        return n - 1;
    }
}
```

Success!

### Refactor VI

Nothing necessarily needed to be done here.

### Red VII

> Perhaps here is where we should begin to implement something a little less generic (and indeed, an *almost* final solution).

Reviewer: We should be thinking about implementation in our green steps rather than red. Red is for breaking things.

Let's check `F(5) = 5` with `checkFib(5, 5)`.

### Green VII

Given the problem at hand and a bit of insight, let's go with the following:

```java
class Fib {
    int fib(int n) {
        if (n < 1) {
            return -1;
        }
        if (n == 1 || n == 2) {
            return 1;
        }
        return fib(n - 1) + fib(n - 2);
    }
}
```

It passes!

### Refactor VII

> Now we check if there is any optimization that can be done.

Reviewer: ```suggestion
Now we check if we can optimise for readability.
```
